File: Arrays and Hashmap/product-of-array-except-self.cpp
/*
* Product Of Array Except Self : https://leetcode.com/problems/product-of-array-except-self/description/
*/
#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int arrLen = nums.size();
        vector<int> prefix(arrLen,1);
        vector<int> suffix(arrLen,1);
        vector<int> res;
        int pProd = 1;
        int sProd = 1;
        for(int i=0; i<arrLen; i++){
            pProd*=nums[i];
            sProd*=nums[arrLen-1-i];
            prefix[i]=pProd;
            suffix[arrLen - 1 - i] = sProd;
        }
        for(int i=0; i<arrLen; i++){
            if(i==0)
                res.push_back(suffix[i+1]);
            else if(i == arrLen - 1)
                res.push_back(prefix[i-1]);
            else
                res.push_back(prefix[i-1] * suffix[i+1]);
        }
        return res;
    }
};

int main(){
    vector<int> nums;
    nums.push_back(1);
    nums.push_back(2);
    nums.push_back(3);
    nums.push_back(4);
    Solution s;
    for(int i:s.productExceptSelf(nums))
        cout<<i;
    return 0;
}

File: Arrays and Hashmap/product-of-array-except-self.py
'''
* Product Of Array Except Self : https://leetcode.com/problems/product-of-array-except-self/description/
'''
class Solution(object):
    def productExceptSelf1(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        pref = [1] * n
        suff = [1] * n
        res = [0] * n

        # Build prefix products
        for i in range(1, n):
            pref[i] = nums[i - 1] * pref[i - 1]

        # Build suffix products
        for i in range(n - 2, -1, -1):
            suff[i] = nums[i + 1] * suff[i + 1]

        # Multiply prefix and suffix for result
        for i in range(n):
            res[i] = pref[i] * suff[i]

        return res

# Example usage:
if __name__ == "__main__":  

    sol = Solution()
    print(sol.productExceptSelf1([1, 2, 3, 4]))  # Output: [24, 12, 8, 6]